relay-server/src/actors/project_cache.rs

```diff
@@ -519,30 +613,90 @@ impl ProjectCacheBroker {
         // do not pass `no_cache`. In case the project is rate limited, we do not want to force
         // a full reload. Fetching must not block the store request.
         project.prefetch(false);
+        project.check_envelope(envelope, context)
+    }
+
+    /// Handles the processing of the provided envelope.
+    fn handle_processing(
+        &mut self,
+        state: Arc<ProjectState>,
+        sampling_state: Option<Arc<ProjectState>>,
+        envelope: Box<Envelope>,
+        envelope_context: EnvelopeContext,
+    ) {
+        let project_key = envelope.meta().public_key();
+        // The `Envelope` and `EnvelopeContext` will be dropped if the `Project::check_envelope()`
+        // function returns any error, which will also be ignored here.
+        let Some(Ok(checked)) = self
+            .projects
+            .get_mut(&project_key)
+            .map(|p| p.check_envelope(envelope, envelope_context)) else { return; };
+
+        if let Some((envelope, envelope_context)) = checked.envelope {
+            let mut process = ProcessEnvelope {
+                envelope,
+                envelope_context,
+                project_state: state.clone(),
+                sampling_project_state: None,
+            };
 
-        project.check_envelope(message.envelope, message.context)
+            if let Some(sampling_state) = sampling_state {
+                if state.organization_id == sampling_state.organization_id {
+                    process.sampling_project_state = Some(sampling_state)
+                }
+            }
+
+            EnvelopeProcessor::from_registry().send(process);
+        }
     }
 
+    /// Checks an incoming envelope and decides either process it immediately or buffer it.
+    ///
+    /// Few conditions are checked here:
+    /// - If there is no dynamic sampling key and the project is already cached, we do straight to
+    /// processing otherwise buffer the envelopes.
+    /// - If the dynamic sampling key is provided and if the root and sampling projects
+    /// are cached - process the envelope, buffer otherwise.
+    ///
+    /// This means if the caches are hot we always process all the incoming envelopes without any
+    /// delay. But in case the project state cannot be fetched, we keep buffering till the state
+    /// is eventually updated.
+    ///
+    /// The flushing of the buffered envelopes happens in `update_state`.
     fn handle_validate_envelope(&mut self, message: ValidateEnvelope) {
-        // Preload the project cache for dynamic sampling in parallel to the main one.
-        if let Some(sampling_key) = utils::get_sampling_key(&message.envelope) {
-            self.get_or_create_project(sampling_key)
-                .prefetch(message.envelope.meta().no_cache());
-        }
+        let ValidateEnvelope { envelope, context } = message;
+
+        // Fetch the project state for our key and make sure it's not invalid.
+        let root_key = envelope.meta().public_key();
```

Comment:
See comment on ambiguity of "root project".

https://github.com/getsentry/relay/blob/e7c862e84d3d3cf76127a2ef80dbb7f4b6235082/relay-server/src/actors/processor.rs#L224